pkg/vectorize/bin/bin.go

```diff
@@ -15,59 +15,109 @@
 package bin
 
 import (
+	"math"
+	"math/big"
 	"math/bits"
+	"strconv"
 
 	"github.com/matrixorigin/matrixone/pkg/container/types"
 	"golang.org/x/exp/constraints"
 )
 
 var (
-	Int8BitLen  = bitLen[int8]
-	Int16BitLen = bitLen[int16]
-	Int32BitLen = bitLen[int32]
-	Int64BitLen = bitLen[int64]
-
-	Uint8BitLen  = bitLen[uint8]
-	Uint16BitLen = bitLen[uint16]
-	Uint32BitLen = bitLen[uint32]
-	Uint64BitLen = bitLen[uint64]
-
-	Float32BitLen = bitLen[float32]
-	Float64BitLen = bitLen[float64]
-
-	Int8ToBinary  = toBinary[int8]
-	Int16ToBinary = toBinary[int16]
-	Int32ToBinary = toBinary[int32]
-	Int64ToBinary = toBinary[int64]
-
-	Uint8ToBinary  = toBinary[uint8]
-	Uint16ToBinary = toBinary[uint16]
-	Uint32ToBinary = toBinary[uint32]
-	Uint64ToBinary = toBinary[uint64]
-
-	Float32ToBinary = toBinary[float32]
-	Float64ToBinary = toBinary[float64]
+	Int8BitLen  = intBitLen[int8]
+	Int16BitLen = intBitLen[int16]
+	Int32BitLen = intBitLen[int32]
+	Int64BitLen = intBitLen[int64]
+
+	Uint8BitLen  = intBitLen[uint8]
+	Uint16BitLen = intBitLen[uint16]
+	Uint32BitLen = intBitLen[uint32]
+	Uint64BitLen = intBitLen[uint64]
+
+	Float32BitLen = floatBitLen[float32]
+	Float64BitLen = floatBitLen[float64]
+
+	Int8ToBinary  = intToBinary[int8]
+	Int16ToBinary = intToBinary[int16]
+	Int32ToBinary = intToBinary[int32]
+	Int64ToBinary = intToBinary[int64]
+
+	Uint8ToBinary  = intToBinary[uint8]
+	Uint16ToBinary = intToBinary[uint16]
+	Uint32ToBinary = intToBinary[uint32]
+	Uint64ToBinary = intToBinary[uint64]
+
+	Float32ToBinary = floatToBinary[float32]
+	Float64ToBinary = floatToBinary[float64]
 )
 
-func bitLen[T constraints.Integer | constraints.Float](xs []T) int64 {
+func intBitLen[T constraints.Integer](xs []T) int64 {
 	var r int64
 	for _, x := range xs {
-		// convert to uint64 to count bits
-		val := uint64(x)
-		if val == 0 {
+		if x == 0 {
+			// special case for zero
+			r += 1
+		} else if x > 0 {
+			r += int64(bits.Len64(uint64(x)))
+		} else {
+			// calc two's complement for negative number to improve the compatibility between arm64 and x86 platforms
+			n := big.NewInt(int64(x))
+			mask := big.NewInt(1)
+			mask.Lsh(mask, 64)
+			n.Add(n, mask)
+
+			r += int64(n.BitLen())
+		}
+	}
+	return r
+}
+
```

Comment:
seems like calculating the result length is quite a lot of cpu time, what if we don't calculate the result length ahead and use the slice's append function instead